pkg/sst/sst.go

```diff
@@ -234,3 +238,119 @@ func getBits(val, i, j uint32) uint32 {
 func isBitSet(val, n uint32) bool {
 	return val&(1<<n) != 0
 }
+
+func setBit(val, n uint32) uint32 {
+	return val | (1 << n)
+}
+
+func clearBit(val, n uint32) uint32 {
+	return val &^ (1 << n)
+}
+
+func setBFStatus(info *SstPackageInfo, status bool) error {
+	rsp, err := sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_GET_TDP_CONTROL, uint32(info.PPCurrentLevel))
+	if err != nil {
+		return fmt.Errorf("failed to read SST status: %v", err)
+	}
+
+	req := clearBit(rsp, 17)
+	if status {
+		req = setBit(rsp, 17)
+	}
+
+	if _, err = sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_SET_TDP_CONTROL, req); err != nil {
+		return fmt.Errorf("failed to enable SST %s: %v", "BF", err)
+	}
+
+	info.BFEnabled = status
+
+	return nil
+}
+
+func adjustScalingMinFromBaseFreq(cpu utils.ID) error {
+	minScalingFreq, err := getCPUScalingMinFreq(cpu)
+	if err != nil {
+		sstlog.Debugf("sst-bf cpu %d set scaling freq error: %v", cpu, err)
+		return err
+	}
+
+	baseFreq, err := getCPUBaseFreq(cpu)
+	if err != nil {
+		sstlog.Debugf("sst-bf cpu %d set base freq error: %v", cpu, err)
+		return err
+	}
+
+	// Boost the scaling_min_freq if the base frequency is higher
+	if minScalingFreq < baseFreq {
+		if err := setCPUScalingMinFreq(cpu, baseFreq); err != nil {
+			sstlog.Debugf("sst-bf cpu %d set min scaling freq error: %v", cpu, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func setScalingMin2CPUInfoMax(info *SstPackageInfo) error {
+	for _, cpu := range info.BFCores.Members() {
+		err := setCPUScalingMin2CPUInfoMaxFreq(cpu)
+		if err != nil {
+			sstlog.Debugf("sst-bf cpu %d set min freq error: %v", cpu, err)
+			return err
+		}
+
+		err = adjustScalingMinFromBaseFreq(cpu)
+		if err != nil {
+			sstlog.Debugf("sst-bf cpu %d adjust scaling freq error: %v", cpu, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// EnableBF enables SST-BF and sets it up properly
+func EnableBF(info *SstPackageInfo) error {
```

Comment:
Hmm, maybe a better api would be the caller just to give package id(s) as the argument here. Don't rely on packageinfo given by the caller, it might be total mumbo-jumbo

Reply:
Well, fetching the package info is rather heavy operation so I would let the caller supply the info so that we can reuse it in the API.

Reply:
I'm not sure if the cost of the operation really matters here, in practice, as this is supposed to be a configuration API that is not supposed to be exercised frequently. It also ensures that you don't get "out-of-sync" with the actual state of the system.

One alternative, instead of `SstPackageInfo`, would be to introduce some kind of `sstHandle` type with no exported variables and which would cover all packages.